Win32.xs

```diff
@@ -1557,6 +1558,127 @@ XS(w32_SetConsoleOutputCP)
     XSRETURN_IV(SetConsoleOutputCP((int)SvIV(ST(0))));
 }
 
+XS(w32_GetProcessPrivileges)
+{
+    dXSARGS;
+    BOOL ret;
+    HV *priv_hv;
+    HANDLE proc_handle, token;
+    char *priv_name = NULL;
+    TOKEN_PRIVILEGES *privs = NULL;
+    DWORD i, pid, priv_name_len = 100, privs_len = 300;
+
+    if (items > 1)
+        Perl_croak(aTHX_ "usage: Win32::GetProcessPrivileges([$pid])");
+
+    if (items == 0) {
+        EXTEND(SP, 1);
+        pid = GetCurrentProcessId();
+    }
+    else {
+        pid = (DWORD)SvUV(ST(0));
+    }
+
+    proc_handle = OpenProcess(PROCESS_QUERY_INFORMATION, FALSE, pid);
+
+    if (!proc_handle)
+        XSRETURN_NO;
+
+    ret = OpenProcessToken(proc_handle, TOKEN_QUERY, &token);
+    CloseHandle(proc_handle);
+
+    if (!ret)
+        XSRETURN_NO;
+
+    do {
+        Renewc(privs, privs_len, char, TOKEN_PRIVILEGES);
+        ret = GetTokenInformation(
+            token, TokenPrivileges, privs, privs_len, &privs_len
+        );
+    } while (!ret && GetLastError() == ERROR_INSUFFICIENT_BUFFER);
+
+    CloseHandle(token);
+
+    if (!ret) {
+        Safefree(privs);
+        XSRETURN_NO;
+    }
+
+    priv_hv = newHV();
+    New(0, priv_name, priv_name_len, char);
+
+    for (i = 0; i < privs->PrivilegeCount; ++i) {
+        DWORD ret_len = 0;
+        LUID_AND_ATTRIBUTES *priv = &privs->Privileges[i];
+        BOOL is_enabled = !!(priv->Attributes & SE_PRIVILEGE_ENABLED);
+
+        if (priv->Attributes & SE_PRIVILEGE_REMOVED)
+            continue;
+
+        do {
+            ret_len = priv_name_len;
+            ret = LookupPrivilegeNameA(
+                NULL, &priv->Luid, priv_name, &ret_len
+            );
+
+            if (ret_len > priv_name_len) {
+                priv_name_len = ret_len + 1;
+                Renew(priv_name, priv_name_len, char);
+            }
+        } while (!ret && GetLastError() == ERROR_INSUFFICIENT_BUFFER);
+
+        if (!ret) {
+            SvREFCNT_dec((SV*)priv_hv);
+            Safefree(privs);
+            Safefree(priv_name);
+            XSRETURN_NO;
+        }
+
+        hv_store(priv_hv, priv_name, ret_len, newSViv(is_enabled), 0);
```

Comment:
You are using `!!0` in other places to get a proper `false` value, but here you store the boolean as an integer. What is the reason you don't use `is_enabled ? &PL_sv_yes : &PL_sv_no`?

I'm fine with merging as-is; just curious.

Reply:
I agree it's inconsistent, I'll change it.

Reply:
On a second thought, those proper booleans are a bit problematic in larger structures. The way how Data::Dumper represents them is very confusing:

```
$VAR1 = {
          'SeTimeZonePrivilege' => '',
          'SeUndockPrivilege' => ${\$VAR1->{'SeTimeZonePrivilege'}},
          'SeChangeNotifyPrivilege' => 1,
          'SeShutdownPrivilege' => ${\$VAR1->{'SeTimeZonePrivilege'}},
          'SeIncreaseWorkingSetPrivilege' => ${\$VAR1->{'SeTimeZonePrivilege'}}
        };
 ```
 
 Also, apparently they make hash keys effectively read-only. I think in this particular case, I'd rather avoid them.

Reply:
Hmm, that's because multiple elements point to the same (immortal) SV. Maybe `newSVsv(&PL_sv_yes)` would be the proper XS invocation.

But as I said, I'm also fine with leaving it as-is... 😄